File: ide/studio/src/main/kotlin/io/devtech/integration/studio/spec/SpecParser.kt
package io.devtech.integration.studio.spec

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.readValue
import io.devtech.integration.studio.model.*
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.PathItem
import io.swagger.v3.parser.OpenAPIV3Parser
import io.swagger.v3.parser.core.models.ParseOptions

class SpecParser {

    private val yamlMapper = ObjectMapper(YAMLFactory())

    fun parse(yamlContent: String): ParsedSpec {
        // Parse raw YAML to extract $ref names before resolution
        val rawDoc = yamlMapper.readValue<Map<String, Any?>>(yamlContent)

        val options = ParseOptions().apply { isResolve = true }
        val result = OpenAPIV3Parser().readContents(yamlContent, null, options)
        val api = result.openAPI
            ?: throw IllegalArgumentException(
                "Failed to parse OpenAPI spec: ${result.messages.joinToString("; ")}"
            )

        return ParsedSpec(
            title = api.info?.title ?: "Untitled",
            operations = extractOperations(api, rawDoc),
            schemas = extractSchemas(api),
        )
    }

    private fun extractOperations(api: OpenAPI, rawDoc: Map<String, Any?>): List<OperationDef> {
        val operations = mutableListOf<OperationDef>()
        val paths = api.paths ?: return operations
        @Suppress("UNCHECKED_CAST")
        val rawPaths = rawDoc["paths"] as? Map<String, Any?> ?: emptyMap()

        for ((path, pathItem) in paths) {
            @Suppress("UNCHECKED_CAST")
            val rawPathItem = rawPaths[path] as? Map<String, Any?> ?: emptyMap()
            val methods = mapOf(
                "get" to pathItem.get,
                "post" to pathItem.post,
                "put" to pathItem.put,
                "patch" to pathItem.patch,
                "delete" to pathItem.delete,
            )

            for ((method, op) in methods) {
                if (op?.operationId == null) continue

                val parameters = mutableListOf<ParameterDef>()
                for (p in (pathItem.parameters ?: emptyList()) + (op.parameters ?: emptyList())) {
                    parameters.add(
                        ParameterDef(
                            name = p.name,
                            `in` = p.`in` ?: "query",
                            required = p.required ?: false,
                            type = p.schema?.type ?: "string",
                        )
                    )
                }

                // Extract $ref names from raw doc
                @Suppress("UNCHECKED_CAST")
                val rawOp = rawPathItem[method] as? Map<String, Any?>
                val requestSchema = extractRequestSchema(rawOp, op)
                val responseSchema = findResponseSchema(rawOp, op)

                operations.add(
                    OperationDef(
                        operationId = op.operationId,
                        method = method.uppercase(),
                        path = path,
                        summary = op.summary,
                        requestSchema = requestSchema,
                        responseSchema = responseSchema,
                        parameters = parameters,
                    )
                )
            }
        }
        return operations
    }

    @Suppress("UNCHECKED_CAST")
    private fun extractRequestSchema(rawOp: Map<String, Any?>?, op: io.swagger.v3.oas.models.Operation): String? {
        // Try raw $ref first
        val rawBody = rawOp?.get("requestBody") as? Map<String, Any?>
        val rawContent = rawBody?.get("content") as? Map<String, Any?>
        val rawJson = rawContent?.get("application/json") as? Map<String, Any?>
        val rawSchema = rawJson?.get("schema") as? Map<String, Any?>
        val ref = rawSchema?.get("\$ref") as? String
        if (ref != null) return ref.split("/").last()

        // Fall back to resolved title
        return op.requestBody?.content?.get("application/json")?.schema?.title
    }

    @Suppress("UNCHECKED_CAST")
    private fun findResponseSchema(rawOp: Map<String, Any?>?, op: io.swagger.v3.oas.models.Operation): String? {
        val rawResponses = rawOp?.get("responses") as? Map<String, Any?>
        // Try raw $ref first, then resolved title
        for (responses in listOfNotNull(rawResponses)) {
            for (code in listOf("200", "201", "2XX", "default")) {
                val resp = responses[code] as? Map<String, Any?> ?: continue
                val content = resp["content"] as? Map<String, Any?> ?: continue
                val json = content["application/json"] as? Map<String, Any?> ?: continue
                val schema = json["schema"] as? Map<String, Any?> ?: continue
                val ref = schema["\$ref"] as? String
                if (ref != null) return ref.split("/").last()
                val items = schema["items"] as? Map<String, Any?>
                val itemsRef = items?.get("\$ref") as? String
                if (itemsRef != null) return itemsRef.split("/").last()
            }
        }
        // Fall back to resolved
        val resolvedResponses = op.responses ?: return null
        for (code in listOf("200", "201", "2XX", "default")) {
            val resp = resolvedResponses[code] ?: continue
            val schema = resp.content?.get("application/json")?.schema ?: continue
            if (schema.title != null) return schema.title
        }
        return null
    }

    private fun extractSchemas(api: OpenAPI): List<SchemaDef> {
        val schemas = mutableListOf<SchemaDef>()
        val components = api.components?.schemas ?: return schemas

        for ((name, schema) in components) {
            val fields = mutableListOf<FieldDef>()
            for ((fieldName, fieldSchema) in (schema.properties ?: emptyMap())) {
                fields.add(
                    FieldDef(
                        name = fieldName,
                        type = fieldSchema.type ?: "string",
                        format = fieldSchema.format,
                        description = fieldSchema.description,
                    )
                )
            }
            schemas.add(
                SchemaDef(
                    name = name,
                    fields = fields,
                    required = schema.required ?: emptyList(),
                )
            )
        }
        return schemas
    }
}

File: ide/bff/src/services/spec-parser.ts
import SwaggerParser from '@apidevtools/swagger-parser'
import type { ParsedSpec, OperationDef, SchemaDef, FieldDef, ParameterDef } from '../types'

export async function parseOpenApiSpec(content: string): Promise<ParsedSpec> {
  const yaml = await import('yaml')
  const doc = yaml.parse(content)

  // Parse raw doc first to extract $ref names before they're resolved
  const raw = structuredClone(doc)
  // Then validate (which also dereferences)
  const api = await SwaggerParser.validate(doc) as any

  const operations: OperationDef[] = []
  const paths = api.paths || {}

  for (const [path, pathItem] of Object.entries(paths) as any[]) {
    const methods = ['get', 'post', 'put', 'patch', 'delete'] as const
    for (const method of methods) {
      const op = pathItem[method]
      if (!op || !op.operationId) continue

      const parameters: ParameterDef[] = [
        ...(pathItem.parameters || []),
        ...(op.parameters || []),
      ].map((p: any) => ({
        name: p.name,
        in: p.in,
        required: p.required || false,
        type: p.schema?.type || 'string',
      }))

      // Extract $ref names from the raw (unresolved) spec
      const rawOp = raw.paths?.[path]?.[method]
      const requestSchema = rawOp?.requestBody
        ?.content?.['application/json']
        ?.schema?.$ref?.split('/')?.pop()
        || op.requestBody?.content?.['application/json']?.schema?.title
        || undefined

      const responseSchema = findResponseSchema(rawOp?.responses, op.responses)

      operations.push({
        operationId: op.operationId,
        method: method.toUpperCase(),
        path,
        summary: op.summary,
        requestSchema,
        responseSchema,
        parameters,
      })
    }
  }

  const schemas: SchemaDef[] = []
  const components = api.components?.schemas || {}
  for (const [name, schema] of Object.entries(components) as any[]) {
    const fields: FieldDef[] = []
    for (const [fieldName, fieldSchema] of Object.entries(schema.properties || {}) as any[]) {
      fields.push({
        name: fieldName,
        type: fieldSchema.type || 'string',
        format: fieldSchema.format,
        description: fieldSchema.description,
      })
    }
    schemas.push({
      name,
      fields,
      required: schema.required || [],
    })
  }

  return {
    title: api.info?.title || 'Untitled',
    operations,
    schemas,
  }
}

function findResponseSchema(rawResponses: any, resolvedResponses: any): string | undefined {
  // Try raw first (has $ref), fall back to resolved (has title)
  for (const responses of [rawResponses, resolvedResponses]) {
    if (!responses) continue
    for (const code of ['200', '201', '2XX', 'default']) {
      const schema = responses[code]?.content?.['application/json']?.schema
      if (schema?.$ref) return schema.$ref.split('/').pop()
      if (schema?.title) return schema.title
      if (schema?.items?.$ref) return schema.items.$ref.split('/').pop()
    }
  }
  return undefined
}
